PS4k/src/com/vitco/export/MagicaVoxelExporter.java

```diff
@@ -0,0 +1,123 @@
+package com.vitco.export;
+
+import com.vitco.core.data.Data;
+import com.vitco.core.data.container.Voxel;
+import com.vitco.layout.content.console.ConsoleInterface;
+import com.vitco.util.components.progressbar.ProgressDialog;
+
+import java.io.File;
+import java.io.IOException;
+import java.awt.Color;
+import com.vitco.util.misc.BiMap;
+
+public class MagicaVoxelExporter extends AbstractExporter {
+
+    private final static int PALETTE_SIZE = 256;
+    private final static int MV_VERSION = 150;
+
+    public MagicaVoxelExporter(File exportTo, Data data, ProgressDialog dialog, ConsoleInterface console) throws IOException {
+        super(exportTo, data, dialog, console);
+    }
+
+    // Write File
+    @Override
+    protected boolean writeFile() throws IOException {
+        final int voxelsCount = data.getVisibleLayerVoxel().length;
+        final int[][] sizeMeta = this.getSizeMeta();
+
+        final BiMap<Integer, Color> palette = new BiMap<>();
+        int currentPaletteIndex = PALETTE_SIZE - 2;
+        this.setColorDefault(palette);
+
+        // Magic number
+        fileOut.writeASCIIString("VOX ");
+        // Write version
+        fileOut.writeIntRev(MV_VERSION);
+
+        //MAIN Chunk
+        fileOut.writeASCIIString("MAIN");
+        fileOut.writeIntRev(0); // Content Size
+        fileOut.writeIntRev((12 + 4 * 3) + (12 + 4 + 4 * voxelsCount) + (12 + 4 * PALETTE_SIZE));
+
+        // Size Chunk
+        fileOut.writeASCIIString("SIZE");
+        fileOut.writeIntRev(4 * 3);
+        fileOut.writeIntRev(0);
+        fileOut.writeIntRev(sizeMeta[1][0] + sizeMeta[0][0]);
+        fileOut.writeIntRev(sizeMeta[1][2] + sizeMeta[0][2]); // Maybe MV is Z-UP, Not sure but this just works
+        fileOut.writeIntRev(sizeMeta[1][1] + sizeMeta[0][1]);
+
+
+        //XYZI Chunk
+        fileOut.writeASCIIString("XYZI");
+        fileOut.writeIntRev(4 + 4 * voxelsCount);
+        fileOut.writeIntRev(0);
+        fileOut.writeIntRev(voxelsCount);
+
+        for (Voxel voxel : data.getVisibleLayerVoxel()) {
+
+            int colorId;
+            if (!palette.containsValue(voxel.getColor())) {
```

Comment:
I would just have a TIntHashSet in AbstractExporter and then a int[] getColors() function to expose it. The BiMap still lives here. We don't really care about duplicates (TIntHashSet is almost no memory)...